src/pybind/mgr/cephadm/schedule.py

```diff
@@ -0,0 +1,183 @@
+import logging
+import random
+from typing import List, Optional, Callable, Iterable, Tuple
+
+import orchestrator
+from ceph.deployment.service_spec import PlacementSpec, HostPlacementSpec, ServiceSpec
+from orchestrator import OrchestratorValidationError
+
+logger = logging.getLogger(__name__)
+
+class BaseScheduler(object):
+    """
+    Base Scheduler Interface
+
+    * requires a placement_spec
+
+    `place(host_pool)` needs to return a List[HostPlacementSpec, ..]
+    """
+
+    def __init__(self, placement_spec):
+        # type: (PlacementSpec) -> None
+        self.placement_spec = placement_spec
+
+    def place(self, host_pool, count=None):
+        # type: (List, Optional[int]) -> List[HostPlacementSpec]
+        raise NotImplementedError
+
+
+class SimpleScheduler(BaseScheduler):
+    """
+    The most simple way to pick/schedule a set of hosts.
+    1) Shuffle the provided host_pool
+    2) Select from list up to :count
+    """
+    def __init__(self, placement_spec):
+        super(SimpleScheduler, self).__init__(placement_spec)
+
+    def place(self, host_pool, count=None):
+        # type: (List, Optional[int]) -> List[HostPlacementSpec]
+        if not host_pool:
+            return []
+        host_pool = [x for x in host_pool]
+        # shuffle for pseudo random selection
+        random.shuffle(host_pool)
+        return host_pool[:count]
+
+
+class HostAssignment(object):
+    """
+    A class to detect if hosts are being passed imperative or declarative
+    If the spec is populated via the `hosts/hosts` field it will not load
+    any hosts into the list.
+    If the spec isn't populated, i.e. when only num or label is present (declarative)
+    it will use the provided `get_host_func` to load it from the inventory.
+
+    Schedulers can be assigned to pick hosts from the pool.
+    """
+
+    def __init__(self,
+                 spec,  # type: ServiceSpec
+                 get_hosts_func,  # type: Callable[[Optional[str]],List[str]]
+                 get_daemons_func, # type: Callable[[str],List[orchestrator.DaemonDescription]]
+
+                 filter_new_host=None, # type: Optional[Callable[[str],bool]]
+                 scheduler=None,  # type: Optional[BaseScheduler]
+                 ):
+        assert spec and get_hosts_func and get_daemons_func
+        self.spec = spec  # type: ServiceSpec
+        self.scheduler = scheduler if scheduler else SimpleScheduler(self.spec.placement)
+        self.get_hosts_func = get_hosts_func
+        self.get_daemons_func = get_daemons_func
+        self.filter_new_host = filter_new_host
+        self.service_name = spec.service_name()
+
+
+    def validate(self):
+        self.spec.validate()
+
+        if self.spec.placement.hosts:
+            explicit_hostnames = {h.hostname for h in self.spec.placement.hosts}
+            unknown_hosts = explicit_hostnames.difference(set(self.get_hosts_func(None)))
```

Comment:
Care to elaborate on the explicit `None` passing to `get_hosts_func`?

Reply:
now depends on #34860